### PR TITLE
menubar.utils.lua - remove "function is_format_supported" by making it obsolete. fixes #2468

### PR DESCRIPTION
**Pull request by Sorky**
Added more extensive test cases

https://github.com/awesomeWM/awesome/pull/2468

### CODE REVIEW

**Comment by Sorky** [comment]
I believe I have managed to combine the code change and the test change and additional test files (actual content doesn't matter - that they exist should be enough)

See https://github.com/awesomeWM/awesome/pull/2468 for original PR that this should replace

File: lib/menubar/utils.lua [hunk 1]
```diff
@@ -226,21 +214,23 @@ function utils.lookup_icon_uncached(icon_file)
         return false
     end
 
-    if icon_file:sub(1, 1) == '/' and is_format_supported(icon_file) then
+    local icon_file_ext = icon_file:match(".*%.(.-)$")
```
Comment (actionless) [inline lib/menubar/utils.lua]: i see the changes proposed by @blueyed to make it faster are not incorporated yet: https://github.com/awesomeWM/awesome/pull/2468#issuecomment-437753145

**Comment by Elv13** [comment]
Thanks a lot for this. Right now the new tests don't pass in the CI. Once that's fixed it will be a great step in the right direction of addressing the last large part of Awesome that isn't well tested/testable.

File: lib/menubar/utils.lua (hunk 1 above)
Comment (Sorky) [reply lib/menubar/utils.lua]: His changes are included...

original: local icon_file_ext = icon_file:gsub(".*%.","")
his: local icon_file_ext = icon_file:match("^.+(%..+)$")
final: local icon_file_ext = icon_file:match(".*%.(.-)$")

why it changed...

gsub  ->  match
:  definitely faster [thanks @blueyed]

match("^.+(%..+)$")  ->  match(".*%.(.-)$")
```
icon_file = "/home/user/.config/awesome/theme/name/.icons/file.ext"
print("1>",icon_file:match("^.+(%..+)$") )
print("2>",icon_file:match(".*%.(.-)$") )
icon_file = "/home/user/.config/awesome/theme/name/.icons/file."
print("3>",icon_file:match("^.+(%..+)$") )
print("4>",icon_file:match(".*%.(.-)$") )
1>	.ext
2>	ext
3>	.icons/file.
4>	
```
:  almost identical but "b>" is what is needed

Comment (actionless) [reply lib/menubar/utils.lua]: ok, thanks for the explanation

**Comment by Sorky** [comment]
Travis CI difficulty...
```
>> assert.matches('?', tostring(file_readable('/usr/share/icon5.png')))

Failure → spec/menubar/utils_spec.lua @ 81
menubar.utils lookup_icon_uncached finds icons even those not in the search paths when full path specified
spec/menubar/utils_spec.lua:90: Expected strings to match.
Passed in:
(string) 'nil'
Expected:
(string) '?'
```
After many adjustments I'm stuck - File should have been added...
https://github.com/awesomeWM/awesome/pull/2542/commits/45760241eab90676bdb95c6a3db9bbcdcaf408da

https://github.com/awesomeWM/awesome/pull/2542/commits/dc1164989280ce866d220b4c07f3d3cb4ef7415b

**Comment by Sorky** [comment]
> > Travis CI difficulty...
> 
> you mean locally tests are passing for you but that happens only on travis?

The test that errors (now commented out) really should work - Works on my tests - Something with Travis?

With all my stuffing around, it would be nice to somehow squash out all the old commits

**Comment by psychon** [comment]
Uhm. Travis basically works.

The short version: Everything should work after fixing the luacheck warnings and squashing some of these 36 commits so that just a few commits remain.

The long version:
One build fails because of this:
```
checking spec/menubar/utils_spec.lua              4 warnings
    spec/menubar/utils_spec.lua:90:121: line is too long (148 > 120)
    spec/menubar/utils_spec.lua:91:121: line is too long (138 > 120)
    spec/menubar/utils_spec.lua:107:121: line is too long (130 > 120)
    spec/menubar/utils_spec.lua:121:121: line is too long (134 > 120)
```
The other build fails because it also tests old commits (but only after the tests for HEAD succeeded) and some older commits do not work:
```
Checks failed for these commits:
106fa11 Update utils_spec.lua (Sorky, Sun Jan 6 16:44:54 2019 +1100)
e56a999 Update utils_spec.lua (Sorky, Sun Jan 6 15:51:10 2019 +1100)
0e8ae26 Update utils_spec.lua (Sorky, Sun Jan 6 17:49:03 2019 +1100)
b92ed4c Update utils_spec.lua (Sorky, Sun Jan 6 15:40:08 2019 +1100)
d9fe135 Update utils_spec.lua (Sorky, Sun Jan 6 17:59:44 2019 +1100)
e4c73d4 Update utils_spec.lua (Sorky, Sun Jan 6 15:25:30 2019 +1100)
2d2718d Update utils_spec.lua (Sorky, Sun Jan 6 18:21:42 2019 +1100)
ae792ca Update utils_spec.lua (Sorky, Sun Jan 6 15:12:58 2019 +1100)
3042ab4 Sorky patch 1 (#7) (Sorky, Sat Jan 5 03:34:28 2019 +1100)
e646cf2 Update utils_spec.lua (Sorky, Sun Jan 6 18:35:26 2019 +1100)
4f4560e Update utils_spec.lua (Sorky, Sun Jan 6 19:19:39 2019 +1100)
ce5a2fe Create icon6.xpm (#6) (Sorky, Sat Jan 5 03:27:06 2019 +1100)
9f34a3e Create icon7.svg (#5) (Sorky, Sat Jan 5 03:24:37 2019 +1100)
b837600 Update utils_spec.lua (Sorky, Sun Jan 6 19:26:46 2019 +1100)
a71b908 Update utils_spec.lua (Sorky, Sun Jan 6 20:17:49 2019 +1100)
ac7f28d Merge pull request #4 from Sorky/patch-10 (Sorky, Sat Jan 5 03:22:31 2019 +1100)
56a4d27 Update icon5.png (Sorky, Sun Jan 6 20:20:15 2019 +1100)
98b2384 Update icon6.xpm (Sorky, Sun Jan 6 20:23:34 2019 +1100)
dc11649 Merge pull request #3 from Sorky/Sorky-patch-1 (Sorky, Sat Jan 5 03:17:48 2019 +1100)
6bb9e77 Update icon7.svg (Sorky, Sun Jan 6 20:24:40 2019 +1100)
3f64d8c Update icon6.xpm (Sorky, Sun Jan 6 20:28:56 2019 +1100)
4aed431 Merge pull request #2 from Sorky/patch-4 (Sorky, Sat Jan 5 01:51:31 2019 +1100)
8ff1e11 Update icon7.svg (Sorky, Sun Jan 6 20:29:06 2019 +1100)
07a2387 Update utils_spec.lua (Sorky, Sun Jan 6 20:39:44 2019 +1100)
02d9062 Update utils_spec.lua (Sorky, Sun Jan 6 20:48:52 2019 +1100)
2955c56 Update utils_spec.lua (Sorky, Sun Jan 6 21:03:23 2019 +1100)
c2052c3 Update utils_spec.lua (Sorky, Fri Jan 4 20:22:41 2019 +1100)
c300ce9 Update utils_spec.lua (Sorky, Sun Jan 6 21:24:13 2019 +1100)
```

**Comment by Sorky** [comment]
> Uhm. Travis basically works.
> 
I meant the test code being rum by Travis

Anyone able to do a squash?

**Comment by Sorky** [comment]
I'd do the squash but it doesn't give me the option...

https://blog.github.com/2016-04-01-squash-your-commits/

What’s changing?

Repository administrators now have a few options to choose from when deciding how to handle history.

New merge button settings

**Allow merge commits and commit squashing**

This option will leave the decision to create a merge commit or squash up to the user doing the merging. This lets repository administrators stay flexible when deciding whether or not to retain all history from a feature branch.

**Only allow merge commits**

This is the default behavior and is exactly how the merge button worked before this change. Collaborators won’t have the option to squash their commits via the merge button.

**Only allow squash commits**

This is a new option which lets you force commit squashing on all pull requests merged via the merge button.

**Comment by actionless** [comment]
search for it on git-scm.com not on blog.github.com

**Comment by Sorky** [comment]
What's the diff? 

BTW = All my commits have been done through the web / never command line

**Comment by actionless** [comment]
that's not a reason do be proud of

try this: https://try.github.io/

**Comment by actionless** [comment]
github web ui is fine when you need to quickly change some typo when you're on the go

**Comment by Sorky** [comment]
Never said I was proud of it - that's just the way I've been doing it - Any advice on doing the squash via github web?

**Comment by blueyed** [comment]
That could be done when merging.
If you want to do it yourself you need to use a local client and then force-push.

**Comment by Elv13** [comment]
> Never said I was proud of it - that's just the way I've been doing it - Any advice on doing the squash via github web?

    git rebase -i origin/master
    # add "s" for squash in each commit
    git push --force Sorky/patch-8

**Comment by Sorky** [comment]
> That could be done when merging.

I was on the lookout for that but I think it may be disabled for the repo - needs an admin to check/enable I believe

**Comment by blueyed** [comment]
@Sorky 
We see this button.. :)
So feel free to just let us squash-merge it in the end.
But doing it manually might be a good practice for you.

**Comment by Sorky** [comment]
> @Sorky
> We see this button.. :)
> So feel free to just let us squash-merge it in the end.
> But doing it manually might be a good practice for you.

Please go ahead - I'll look at doing a PR via git next time round

File: spec/menubar/utils_spec.lua [hunk 2]
```diff
@@ -65,6 +65,64 @@ describe("menubar.utils lookup_icon_uncached", function()
...
+
+        -- Shimmed icon base directories contain the following icons:
+        --
+        --     .icons/icon6.xpm
+        --     .icons/icon7.svg
+
+        assert_found_in_path('icon6', '/.icons/icon6.xpm') -- Similar to original tests and testing xpm extension
+        assert_found_in_path('icon7', '/.icons/icon7.svg') -- Similar to original tests and testing svg extension
+
+        assert.is_false(utils.lookup_icon_uncached('/png')) -- [For Uli] supported file does not exist in given location
```
Comment (actionless) [inline spec/menubar/utils_spec.lua]: `[For Uli]`?

Comment (Elv13) [reply spec/menubar/utils_spec.lua]: Please clean this up, people reading this in the deep future will be confused. After that I will approve/merge. You can use `git commit -a --amend --no-edit && git push --force <origin_name> <branch_name>` to update the commit without creating tons of little commits.

**Comment by Elv13** [comment]
I removed all the little temporary commit messages that were squased and squashed the newest 3 commits

File: spec/menubar/utils_spec.lua [hunk 3]
```diff
@@ -65,6 +65,65 @@ describe("menubar.utils lookup_icon_uncached", function()
...
+-- Check for icon not in search path :: if icon_file:sub(1, 1) == '/' and supported_icon_formats[icon_file_ext] then
+    it('finds icons even those not in the search paths when full path specified', function()
+
+        -- Shimmed icon base directories contain the following icons:
+        --
+        --     usr/share/icon5.png
+        --     usr/share/icon6.xpm
+        --     usr/share/icon7.svg
+
+--        assert_found_in_path('/usr/share/icon5.png', '/usr/share/icon5.png')
```
Comment (actionless) [inline spec/menubar/utils_spec.lua]: why those are commented out?

Comment (Sorky) [reply spec/menubar/utils_spec.lua]: As per my previous comments regarding local test cases that work but don't under Travis...
`ToDo?: 3 test lines to restore & resolve why they fail`

**Comment by Sorky** [comment]
> I removed all the little temporary commit messages that were squased and squashed the newest 3 commits

Thanks Elv13 - And apologies

File: spec/menubar/utils_spec.lua (hunk 3 above)
Comment (Sorky) [reply spec/menubar/utils_spec.lua]: My local tests (files exist)...
```
local utils = require("menubar.utils")
print("1>",utils.lookup_icon_uncached('/usr/share/icon5.png'))
print("2>",utils.lookup_icon_uncached('/usr/share/icon6.xpm'))
print("3>",utils.lookup_icon_uncached('/usr/share/icon7.svg'))
1>	/usr/share/icon5.png
2>	/usr/share/icon6.xpm
3>	/usr/share/icon7.svg
```

Comment (actionless) [reply spec/menubar/utils_spec.lua]: yeah, that makes sense, there are no such icons under those filepaths like `/usr/share/icon5.png`

see setUp function of this test module to understand how filepath thing is mocked here

Comment (Sorky) [reply spec/menubar/utils_spec.lua]: Part of the PR adds the files

These other files that were added as part of the PR work...
```
        assert_found_in_path('icon6.xpm', '/.icons/icon6.xpm')
        assert_found_in_path('icon7.svg', '/.icons/icon7.svg')

```

Comment (actionless) [reply spec/menubar/utils_spec.lua]: please read setup function aloud line by line and let me know if you still didn't understood something

Comment (Sorky) [reply spec/menubar/utils_spec.lua]: PS: The path is already in the setup() function and the files are added as part of the PR         ```
shim('get_system_data_dirs', {
            root .. "/usr/local/share",
            root .. "/usr/share"
})
```

Comment (Sorky) [reply spec/menubar/utils_spec.lua]: This where my understanding of Travis is limited. Doesn't isn't `local root = (os.getenv("SOURCE_DIRECTORY") or '.') .. "/spec/menubar'` and `root .. "/usr/share"`allow Travis to simulate  the '/usr/share' directory?

Please note that these work...
```
        assert_found_in_path('icon1', '/usr/share/pixmaps/icon1.png')
        assert_found_in_path('icon2', '/usr/share/icons/icon2.png')
```
As does...
`        assert_found_in_path('icon5.png', '/.icons/icon5.png')
`
Could you possibly copy icon5.png into https://github.com/awesomeWM/awesome/tree/master/spec/menubar/usr/share just to see if Travis is having trouble?

Comment (actionless) [reply spec/menubar/utils_spec.lua]: travis is nothing to do with it, it seems you just have those files in /usr/share locally

Comment (Sorky) [reply spec/menubar/utils_spec.lua]: I added them as part of the PR...

clipboard01

Comment (actionless) [reply spec/menubar/utils_spec.lua]: https://github.com/awesomeWM/awesome/pull/2542#discussion_r247314282
https://github.com/awesomeWM/awesome/pull/2542#discussion_r247314340

Comment (Sorky) [reply spec/menubar/utils_spec.lua]: Solved

**Comment by Sorky** [comment]
Sorry @blueyed and @Elv13 and @actionless but I had to recreate this PR due to "unknown repository"

https://github.com/awesomeWM/awesome/pull/2601

**Pull request closed without merging**